### PR TITLE
Package rewrite - convert Client to an Interface

### LINKED ISSUE #1: SetHeaders is broken
Likely needs to use CanonicalHeader name when setting values in dictionary.

### PR DESCRIPTION
**Pull request by marz619**
Changes the exposed `Client` and returns it as an interface rather than a struct type. Overall, this is an attempt at cleaning up the interface and only exposing the barebones in an attempt to maintain a level of consistency and _future-proofing_ for projects that are using this package.

## Changes

Each subsequent commit is fairly well encapsulated and can be looked at individually to see the overall changes being made.

* update the License
* update the README
* public `Client` is now an `interface` that will implement the necessary  HTTP verb methods for use by the caller. This replicates the current working API that was previously exported as a `struct{}` type.  This interface is implemented internally with a `type client struct` that holds all the configuration options and a reference to the underlying `*http.Client`.

```go
type Client interface {
	// Client returns the underlying *http.Client
	Client() *http.Client

	// Close all Idle connections
	Close()

	// HTTP Request Methods //

	// Do is the generic HTTP request method. The two string parameters in
	// order are the HTTP Method, and the URL to request respectively
	Do(ctx context.Context, rh ResponseHandler, method, url string, body io.Reader, opts ...RequestOption) error

	// GET method
	Get(ctx context.Context, rh ResponseHandler, url string, opts ...RequestOption) error

	// POST method
	Post(ctx context.Context, rh ResponseHandler, url string, body io.Reader, opts ...RequestOption) error
}
```

* introduce the `Option` type and move all Client configuration option functions to their own file `client_option.go`
* consolidate all the request methods into the same file `request.go`
* running golint, go vet and updating/adding tests where necessary
* Adding a `go.mod` file
* An attempt (seemingly successful) to fix #1 - 12db943

### CODE REVIEW

File: client_option.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// ResponseHeaderTimeout is a configuration option to pass to client. It limits
+// the time spent reading the headers of the response.
+func ResponseHeaderTimeout(t time.Duration) Option {
+	return func(c *client) error {
+		c.responseHeaderTimeout = t
+		return nil
+	}
+}
+
+// RoundTripperFunc is like http.HandlerFunc, but for RoundTripper interface.
```
Comment (marz619) [inline client_option.go:149]:
```suggestion
// RoundTripperFunc is like http.HandlerFunc for http.RoundTripper interface.
```

File: client.go [hunk 2]
```diff
@@ -13,244 +13,123 @@ import (
...
 	// DefaultKeepAliveTimeout - when socket keep alive check will be performed
 	DefaultKeepAliveTimeout = 90 * time.Second
-	defaultLogPrefix        = "httpclient "
+
+	// log-prefix
+	defaultLogPrefix = "[httpclient]: "
 )
 
-//ErrInvalidValue signifies that an invaild value was given to configartion option
+// ErrInvalidValue signifies that an invaild value was given to configartion option
```
Comment (rafikdraoui) [inline client.go]:
```suggestion
// ErrInvalidValue signifies that an invalid value was given to configuration option
```

File: client.go [hunk 3]
```diff
@@ -13,244 +13,123 @@ import (
...
+// return a new *client with default values
+func newDefaultClient() *client {
+	c := new(client)
+	c.setDefaults()
+	return c
 }
 
-func (c *Client) setDefaults() {
-	c.headers = make(http.Header)
+// set *sensible* default values on the *client
```
Comment (rafikdraoui) [inline client.go]:
```suggestion
// set sensible default values on the *client
```

The `*` can be a bit confusing when it's also used to denote a pointer type.

File: request_option_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+			t.Errorf("trouble when creating the client: %v", err)
+		}
+
+		// startup a server with handler to check that received request headers
+		// are correct
+		s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// add X-Test if we're not already checking for it
+			if _, ok := tc.exp["X-Test"]; !ok {
+				tc.exp["X-Test"] = headers.Values("X-Test")
+			}
```
Comment (rafikdraoui) [inline request_option_test.go]: When I comment out this section, the tests still pass. So either it isn't needed, or else we are not truly exercising some cases.

File: request_option_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		extra := make(http.Header)
+		for k, vs := range tc.exp {
+			for _, v := range vs {
+				extra.Add(k, v)
+			}
+		}
+		// make the request
+		err = c.Get(context.Background(), NoopResponseHandler, s.URL, SetHeaders(extra))
+		if err != nil {
+			t.Errorf("trouble with set headers roundtripper unexpected -> %v", err)
```
Comment (rafikdraoui) [inline request_option_test.go]:
```suggestion
			t.Errorf("unexpected error with SetHeaders: %v", err)
```

File: request_option.go [hunk 6]
```diff
@@ -30,15 +31,39 @@ func DelHeaders(headers http.Header) RequestOption {
...
 	return func(req *http.Request) error {
-		for k, v := range headers {
-			var val []string
-			copy(val, v)
-			req.Header[k] = val
+		for k, vs := range headers {
+			req.Header.Del(k)
+			for i := range vs {
+				req.Header.Add(k, vs[i])
+			}
```
Comment (rafikdraoui) [inline request_option.go:41]: 👍 I think that should do the trick: reading the docs, it seems that accessing the Header map directly bypasses canonicalization of key names, so using the `Add`/`Set`/`Del`/etc. methods should take care of the issue.

I'm assuming that this was the issue: the original bug report is a bit vague ("SetHeader is broken") 😅

**Review by rafikdraoui — COMMENTED** [review]
The code looks good to me. I think the next step is to update some projects to use this branch as a pre-release version, and see how things go!

File: request_option_test.go (hunk 4 above)
Comment (marz619) [reply request_option_test.go]: made the Test more explicit by adding multiple maps to denote exactly what we're adding (extra), and what's expected here: b899ea3

File: request_option_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+				}
+			}
+			// make the request
+			err = c.Get(context.Background(), NoopResponseHandler, s.URL, SetHeaders(extra))
+			if err != nil {
+				t.Errorf("unexpected error with SetHeaders: %v", err)
+			}
+			// close client & server
+			c.Close()
+			s.Close()
```
Comment (rafikdraoui) [inline request_option_test.go]: These could be deferred when `c` and `s` are defined instead. 

I tried making a GitHub suggestion for the change, but it couldn't compute a nice clean diff for it, so here's a patch generated from the git CLI instead!

```diff
diff --git a/request_option_test.go b/request_option_test.go
index ff56f83..9bb24ec 100644
--- a/request_option_test.go
+++ b/request_option_test.go
@@ -66,6 +66,7 @@ func TestSetHeaders(t *testing.T) {
 			if err != nil {
 				t.Errorf("trouble when creating the client: %v", err)
 			}
+			defer c.Close()
 
 			// startup a server with handler to check that received request headers
 			// are correct
@@ -79,6 +80,7 @@ func TestSetHeaders(t *testing.T) {
 					}
 				}
 			}))
+			defer s.Close()
 
 			// create our extra headers
 			extra := make(http.Header)
@@ -92,9 +94,6 @@ func TestSetHeaders(t *testing.T) {
 			if err != nil {
 				t.Errorf("unexpected error with SetHeaders: %v", err)
 			}
-			// close client & server
-			c.Close()
-			s.Close()
 		})
 	}
 }
```

**Pull request merged**